xonsh/tools.py

```diff
@@ -1648,34 +1650,36 @@ def lazyget(self, key, default=None):
         return self._cmds_cache.get(key, default)
 
     def locate_binary(self, name):
-        """Locates an executable on the file system."""
+        """Locates an executable on the file system using the cache."""
+        possibilities = self.get_possible_names(name)
 
         if ON_WINDOWS:
+            from xonsh.dirstack import _get_cwd  # circular dependency
             # Windows users expect to be able to execute files in the same
             # directory without `./`
             cwd = _get_cwd()
 
-            local_bins = [full_name for full_name in self.get_possible_names(name)
-                          if os.path.isfile(full_name)]
-            if local_bins:
-                return os.path.abspath(os.path.relpath(local_bins[0], cwd))
+            local_bin = next((
+                full_name for full_name in possibilities
+                if os.path.isfile(full_name)
+            ), None)
 
-        possibilities = self.get_possible_names(name)
-        matches = possibilities & self.all_commands.keys()
+            if local_bin:
+                return os.path.abspath(os.path.relpath(local_bin, cwd))
 
-        if matches:
-            return self.all_commands[matches.pop()][0]
+        cmds = self.all_commands
+        cached = next((cmd for cmd in possibilities if cmd in cmds), None)
+        if cached:
+            return cmds[cached][0]
         elif os.path.isfile(name) and name != os.path.basename(name):
             return name
 
     def lazy_locate_binary(self, name):
```

Comment:
Should `lazy_locate_binary` not also look in the current directory, similar to what `locate_binary` does. I they share so much code that you could potentially call `lazy_locate_binary()` from `locate_binary()`. With the addition that `locate_binary()` should just update the command cache first.